website/blog/2025/05-05-arbiter-type-class/index.md

```diff
@@ -0,0 +1,280 @@
+    arbitrable.grant(client, grant)
+  }
+}
+```
+
+This makes instantiating the arbiter a hair nicer:
+
+```scala
+val clients1 = IO(Vec(4, Flipped(new ArbiterClient)))
+val arbiter1 = Module(new GenericPriorityArbiter(4, new ArbiterClient, new ArbiterClientArbitrable))
+arbiter1.clients :<>= clients1
+
+val clients2 = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter2 = Module(new GenericPriorityArbiter(4, Decoupled(UInt(8.W)), new DecoupledArbitrable[UInt]))
+arbiter2.clients :<>= clients2
+```
+
+At least we aren't repeating logic anymore, instead we get to just refer to the type class instance.
+
+However, we can do even better.
+
+## Implicit Type Class Instances
+
+Scala has a powerful feature called **implicit resolution**.
+This allows us to avoid passing around the type class instance explicitly.
+Instead, we can define the type class instance as an implicit value and the compiler will automatically find it for us.
+
+Let us rewrite our typeclass instances as implicit values:
+
+```scala
+implicit val arbiterClientArbitrable: Arbitrable[ArbiterClient] =
+  new Arbitrable[ArbiterClient] {
+    def request(a: ArbiterClient) = a.request
+    def grant(a: ArbiterClient, value: Bool) = a.grant := value
+  }
+
+// In chisel3.util, the type is DecoupledIO while we construct instances of it with Decoupled.
+// Note that this is a def because DecoupledIO itself takes a type parameter.
+implicit def decoupledArbitrable[T <: Data]: Arbitrable[DecoupledIO[T]] =
+  new Arbitrable[DecoupledIO[T]] {
+    def request(a: DecoupledIO[T]) = a.valid
+    def grant(a: DecoupledIO[T], value: Bool) = a.ready := value
+  }
+```
+
+And then we can refactor the arbiter to use the implicit type class instance:
+
+```scala
+class GenericPriorityArbiter[A <: Data](nClients: Int, clientType: A)(implicit arbitrable: Arbitrable[A]) extends Module {
+  val clients = IO(Vec(nClients, Flipped(clientType)))
+
+  val requests = clients.map(arbitrable.request(_))
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    arbitrable.grant(client, grant)
+  }
+}
+```
+
+Now, we can instantiate the arbiter without passing the type class instance explicitly:
+
+```scala
+val clients1 = IO(Vec(4, Flipped(new ArbiterClient)))
+val arbiter1 = Module(new GenericPriorityArbiter(4, new ArbiterClient))
+arbiter1.clients :<>= clients1
+
+val clients2 = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter2 = Module(new GenericPriorityArbiter(4, Decoupled(UInt(8.W))))
+arbiter2.clients :<>= clients2
+```
+
+This is much cleaner and more readable.
+
+Scala also has special syntax for implicit typeclass instances:
+
+```scala
+class GenericPriorityArbiter[A <: Data : Arbitrable](nClients: Int, clientType: A) extends Module {
+  ...
+}
+```
+
+This is equivalent to the previous definition, but is more concise.
+Note that unlike the version with the implicit argument, this one does not bind a variable name for the implicit argument.
```

Comment:
is this really relevant for the blog post? is this something people are going to need to do?